docs/engineering/architecture/permissions.md

```diff
@@ -0,0 +1,86 @@
+
+## Adding Connections (backend perspective)
+
+Regardless of UI, the backend should receive a `POST` request to the new RPC endpoint defining a new connection. We'll use the same functions set up in [PR \#3348](https://github.com/mathesar-foundation/mathesar/pull/3348). The relevant info should be stored in the `Database`, `DatabaseServer`, and `DatabaseServerCredential` models. Also, if the user does not already have an entry defining their role on the given database, we could create such an entry automatically in the `UserDatabaseMap` model. This is optional, and doesn't really affect the architecture. Then, to let some other users access that connection, we will provide an RPC function that lets an admin set different users' credentials for the given database by creating or updating `UserDatabaseMap` resources. Note that this does _not_ directly modify anything to do with permissions on actual database objects (e.g., schemata or tables).
+
+## Granting database object privileges
+
+The backend will provide RPC functions that let an admin (who has access to a sufficiently-privileged database role via a connection) access database-level permission granting functionality directly. So, to grant access to create tables in a schema to a Mathesar user, the admin uses an RPC function that defines a privilege-granting query (via a PL/pgSQL function) on the database. Note that this request doesn't actually modify any model instance.
+
+- Privileges on DB objects are thus not granted directly to Mathesar users.
+- Privileges on DB objects are granted to DB roles, which may be accessible to some (or all) Mathesar users.
+- Anytime an RPC function requiring DB access runs, it runs using the connection defined via the `UserDatabaseMap`, with the associated privileges on that database.
+
+!!! question "UX Question"
+    Should the admin think in terms of groups of Mathesar users, or specifically in terms of connections when dealing with DB-level privileges?
+    
+!!! danger "Potential Confusion"
+    In case the admin or DBA want multiple Mathesar users to be able to modify various DB objects, three options are available:
+    - Give all relevant Mathesar users access to connect as the owning DB user.
+    - `GRANT` the owning DB role (ostensibly a user) to DB users connectable by the relevant Mathesar users.
+    - Have at least one DB superuser available for use with a connection, and give relevant Mathesar users access to that DB superuser.
+    
+## Metadata object privileges
+
+Examples of these are table properties like preview columns. Permissions for these are derived from permissions on relevant User DB objects. So, if a user wants to modify the preview template for a table, they would be checked against the permission to modify the underlying table structure. Thus, modifying this metadata will occur in a request to modify the associated table, and the privilege check will be whether the relevant DB user is allowed to `ALTER` that table. 
```

Comment:
This needs to be updated based on our last discussion where we decided that all metadata permissions will now be identified using the `metadata\_role` column in `UserDBRoleMap `.

Reply:
Good catch, you're right